feat(textual): `Coin` and `Coins` value renderers

amaurym opened the pull request:
<!--
The default pull request template is for types feat, fix, or refactor.
For other templates, add one of the following parameters to the url:
- template=docs.md
- template=other.md
-->

## Description

Closes: #12708

Initially was part of #12088, but carved out of that PR into its own.

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

JimLarson commented on tx/textual/valuerenderer/coins.go:
```diff
@@ -0,0 +1,171 @@
+	// If we didn't find either exponent, then we return early.
+	if !foundCoinExp || !foundDispExp {
+		vr, err := formatInteger(coin.Amount)
+		return vr + " " + coin.Denom, err
```
Must ensure that `coin.Denom` is printable ASCII.

amaurym replied:
Related: https://github.com/cosmos/cosmos-sdk/issues/12026

I added the issue to the current SDK sprint. Once we merge that, then a simple comment in this code linking to it is IMO enough. Though can still add a simple check here for extra safety if you think it's useful.

aaronc replied:
Re: https://github.com/cosmos/cosmos-sdk/pull/12729/files#r963716452, this is another area where we may need a slightly different implementation from `sdk.Coin` and/or need to change the way denom validation works generally to ensure bijectivity.

JimLarson commented on tx/textual/valuerenderer/coins.go:
```diff
@@ -0,0 +1,171 @@
+		dispAmount = dispAmount.Quo(math.LegacyNewDec(10).Power(uint64(-exponentDiff)))
+	}
+
+	vr, err := formatDecimal(dispAmount.String())
```
Must ensure `dispDenom` is printable ASCII.

JimLarson commented on tx/textual/valuerenderer/coins.go:
```diff
@@ -0,0 +1,171 @@
+		err        error
+	)
+	if exponentDiff > 0 {
+		dispAmount, err = math.LegacyNewDecFromStr(coin.Amount)
+		if err != nil {
+			return "", err
+		}
```
Hoist these common statements out of the branches.

JimLarson replied:
This is still un-addressed.

JimLarson commented on tx/textual/internal/testdata/coin.json:
```diff
@@ -0,0 +1,26 @@
+  [{"amount": "100000", "denom": "point"}, { "denom": "POINT", "exponent": 0 }, "100'000 POINT"],
+  [{"amount": "1000000", "denom": "point"}, { "denom": "POINT", "exponent": 0 }, "1'000'000 POINT"],
+  [{"amount": "10000000", "denom": "point"}, { "denom": "POINT", "exponent": 0 }, "10'000'000 POINT"]
+]
```
Add a test cases for

- zero amount
- coin without metadata
- metadata with entries other than display and coin names
- metadata without the display or coin name
- failures (see `timestamp.json` for examples)

amaurym replied:
Added test cases for all items except "metadata with entries other than display and coin names". That would be a test-related json parsing error, which I don't think is really useful here?

amaurym replied:
OK, I misunderstood, by `metadata with entries...` I understood fields in the metadata other than the usual `display` etc fields. I added.

JimLarson commented on tx/textual/internal/testdata/coins.json:
```diff
@@ -0,0 +1,52 @@
+      "uxc1": { "denom": "xC1", "exponent": 6 }
+    },
+    "0.000001 xA, 0.000002 xB, 0.000003 xC, 0.000004 xC1"
+  ]
```
Add test cases for an empty `Coins` and a `Coins` of a single `Coin`.

Also add cases with a failure in coins at the beginning, middle, and end or the coins array.

JimLarson replied:
Still needs empty coins and singleton coins.

amaurym replied:
Added.

JimLarson commented on tx/textual/valuerenderer/coins.go:
```diff
@@ -0,0 +1,171 @@
+
+type coinsValueRenderer struct {
+	// coinMetadataQuerier defines a function to query the coin metadata from
+	// state.
```
Add a note that this function is provided for testing convenience - in production always use the standard query function.

amaurym replied:
This is not only used for testing. In production, this should be set to the bank keeper's query function.

JimLarson commented on tx/textual/valuerenderer/coin_test.go:
```diff
@@ -0,0 +1,75 @@
+		require.NoError(t, err)
+		b := new(strings.Builder)
+		err = r.Format(ctx, protoreflect.ValueOf(tc.coin.ProtoReflect()), b)
+		require.NoError(t, err)
```
Add test cases for each error exit from the renderer so you get 100% coverage.

Also use the parse method (which should be implemented for this PR) to prove round-tripping.

amaurym replied:
Added test cases to bring file coverage to 89%. However 100% coverage seems like a futile exercice, lines like [this one](https://github.com/cosmos/cosmos-sdk/blob/4670093c567959ed3928b1c4776dea642c6bf26f/tx/textual/valuerenderer/coins.go#L129) are not worth the effort to cover.

Added an inline comment about parsing, see https://github.com/cosmos/cosmos-sdk/pull/12729#discussion_r962963732

JimLarson commented on tx/textual/valuerenderer/valuerenderer.go:
```diff
@@ -1,25 +1,43 @@
-func NewTextual() Textual {
-	t := Textual{}
+func NewTextual(q CoinMetadataQueryFn) Textual {
+	t := Textual{coinMetadataQuerier: q}
```
You should provide a default querier within this package and add it to the `Textual` here, just like we add all the built-in renderers. Consider adding a `NewRawTextual()` with initialization methods for use by tests and custom rendering uses.

amaurym replied:
What would the default querier be?

amaurym commented on tx/textual/valuerenderer/coins.go:
```diff
@@ -0,0 +1,171 @@
+// formatCoin formats a sdk.Coin into a value-rendered string, using the
+// given metadata about the denom. It returns the formatted coin string, the
+// display denom, and an optional error.
+func formatCoin(coin *basev1beta1.Coin, metadata *bankv1beta1.Metadata) (string, error) {
```
The reason is #11799: the `tx` module is a standalone go module, so it cannot depend on the root module. 

I'll discuss with the SDK team on moving sdk.Coins into another module, probably `core`.

aaronc replied:
I actually don't think it makes sense to bring in all of the existing `sdk.Coin` stuff into core. I commented in #13155. The implementation for `SIGN_MODE_TEXTUAL` will actually be slightly different from what's in `sdk.Coin`. We need bijectivity here and `ParseDecCoin` depends on the global denom regex which is actually too restrictive for a correct implementation for `SIGN_MODE_TEXTUAL` (besides being a mutable global variable). And we will need to do an extra round of allocation if we use `sdk.Coin` so that approach will be less performant (though likely not in a meaningful way I'll admit).

amaurym commented on tx/textual/valuerenderer/coins.go:
```diff
@@ -0,0 +1,171 @@
+}
+
+func (vr coinsValueRenderer) Parse(_ context.Context, r io.Reader) (protoreflect.Value, error) {
+	panic("implement me")
```
I started adding it to this PR, but there's a big scope creek. We need to add `formatIntegers` too, and ideally test cases for that. Instead, I referenced https://github.com/cosmos/cosmos-sdk/issues/13153 everywhere related to parsing, and it's also being tracked in the epic #11970.

I agree it's important, but maybe someone else can pick it up. Is that fine by you @JimLarson ?

JimLarson replied:
If we're still requiring bijectivity, then I'd hardly call implementing bijectivity "scope creep".  I'm fine if you want to split this over two PRs, but don't close #12708 until the second one lands.

JimLarson replied:
Hmm, I see that parsing coins and numbers has been split into #13153, which has been added to the epic. Fine, but I still regard that as essential for the epic, not a nice-to-have.

amaurym replied:
Yes, the epic's phase 1 was actually only meant for `Format` which is the shortest path to get something e2e-testable. But obviously the epic won't be closed as long as all `Parse` are also implemented (i.e. phase 4 done).

amaurym commented on tx/textual/valuerenderer/coins.go:
```diff
@@ -0,0 +1,171 @@
+
+// formatCoin formats a sdk.Coin into a value-rendered string, using the
+// given metadata about the denom. It returns the formatted coin string, the
+// display denom, and an optional error.
```
See https://github.com/cosmos/cosmos-sdk/pull/12729#discussion_r962963732

amaurym commented on tx/go.mod:
```diff
@@ -13,9 +13,15 @@ require (
+	golang.org/x/sys v0.0.0-20220811171246-fbc7d0a398ab // indirect
+	golang.org/x/text v0.3.7 // indirect
+	google.golang.org/genproto v0.0.0-20220519153652-3a47de7e79bd // indirect
+	google.golang.org/grpc v1.48.0 // indirect
```
Unfortunately, I can't run tests without these changes. I simply ran go mod tidy.

amaurym commented on tx/textual/internal/testdata/coin.json:
```diff
@@ -0,0 +1,297 @@
+    "proto": {"amount": "100000", "denom": "ucosm"},
+    "metadata": {"display": "COSM", "denom_units": [{"denom": "COSM", "exponent": 6}, {"denom": "ucosm", "exponent": 0}]},
+    "text": "0.1 COSM"},
+  {
+    "proto": {"amount": "1000000", "denom": "ucosm"},
+    "metadata": {"display": "COSM", "denom_units": [{"denom": "COSM", "exponent": 6}, {"denom": "ucosm", "exponent": 0}]},
+    "text": "1 COSM"},
+  {
+    "proto": {"amount": "10000000", "denom": "ucosm"},
+    "metadata": {"display": "COSM", "denom_units": [{"denom": "COSM", "exponent": 6}, {"denom": "ucosm", "exponent": 0}]},
+    "text": "10 COSM"
+  },
+  {
+    "proto": {"amount": "0", "denom": "COSM"},
+    "metadata": {"display": "ucosm", "denom_units": [{"denom": "COSM", "exponent": 6}, {"denom": "ucosm", "exponent": 0}]},
+    "text": "0 ucosm"
+  },
+  {
```
where exactly?

facundomedica replied:
Some of the closures are in a new line and some are in-line:

```
  },
  {
    "proto": {"amount": "10000000", "denom": "ucosm"},
    "metadata": {"display": "COSM", "denom_units": [{"denom": "COSM", "exponent": 2}]},
    "text": "10'000'000 ucosm"
  },
  {
    "proto": {"amount": "0", "denom": "ucosm"},
    "metadata": {"display": "COSM", "denom_units": [{"denom": "COSM", "exponent": 2}, {"denom": "ucosm", "exponent": 0}]},
    "text": "0 COSM"
  },
  {
    "proto": {"amount": "1", "denom": "ucosm"},
    "metadata": {"display": "COSM", "denom_units": [{"denom": "COSM", "exponent": 2}, {"denom": "ucosm", "exponent": 0}]},
    "text": "0.01 COSM"},
  {
    "proto": {"amount": "10", "denom": "ucosm"},
    "metadata": {"display": "COSM", "denom_units": [{"denom": "COSM", "exponent": 2}, {"denom": "ucosm", "exponent": 0}]},
    "text": "0.1 COSM"},
  {
  ```

amaurym commented:
I'm putting automerge on, to unblock some auto-generated CLI work which needs the coin formatting function.

@JimLarson I believe I addressed all your comments, but can do a follow-up if there's something else left.
